packages/types/config/module.d.ts

```diff
@@ -17,10 +19,35 @@ interface ExtendFunctionContext {
   loaders: NuxtOptionsLoaders
 }
 
-type ExtendFunction = (config: WebpackConfiguration, ctx: ExtendFunctionContext) => void
+export interface ModuleTemplateConfig {
+  src: string,
+  fileName?: string,
+  options?: any
+}
+
+export interface ModuleTemplateDest {
+  src: string
+  dst: string
+  options?: any
+}
+
+export type ModuleTemplate = ModuleTemplateConfig | string
+
+type ExtendRoutesFunction = (routes: NuxtRouteConfig[], resolve: (...pathSegments: string[]) => string) => void
+type ExtendBuildFunction = (config: WebpackConfiguration, ctx: ExtendFunctionContext) => void
 
 interface ModuleThis {
-  extendBuild(fn: ExtendFunction): void
+  addTemplate(template: ModuleTemplate): ModuleTemplateDest
+  addPlugin(template: ModuleTemplate): void
```

Comment:
We need an interface extending `ModuleTemplate` for plugins with `ssr?: boolean` and `mode?: 'all' | 'client' | 'server'`